crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -574,9 +594,29 @@ where
             }
             ast::Expr::Lambda(lambda) => {
                 if let Some(parameters) = &lambda.parameters {
+                    // The default value of the parameters needs to be evaluated in the
+                    // enclosing scope.
+                    for default in parameters
+                        .iter_non_variadic_params()
+                        .filter_map(|param| param.default.as_deref())
+                    {
+                        self.visit_expr(default);
+                    }
                     self.visit_parameters(parameters);
```

Comment:
This `visit_parameters()` call here is confusing me.
- Why do we call `self.visit_parameters()` for a `Lambda` expression, but not for a `FunctionDef` statement?
- Is it correct for the `visit_parameters()` call to take place before the `self.push_scope()` call? Exactly which scope annotations should be evaluated in is a complicated question because it's affected by `from __future__ import annotations` or whether it's a stub file, so this might be correct. But if it is correct, then why can't the default values just be visited as part of the `visit_parameters()` call?
- If we keep visiting the default values outside the `visit_parameters()` call, what would `visit_parameters()` do for a lambda function anyway? Lambda functions can't have annotations on their parameters, so for lambdas the only things we need to analyse are the binding the parameter creates and the parameter's default value (if it has one). Both the binding and the default value seem to be taken care of outside the `visit_parameters()` call

Reply:
I think @carljm answered to most of @AlexWaygood questions (thanks Carl).

> * If we keep visiting the default values outside the `visit_parameters()` call, what would `visit_parameters()` do for a lambda function anyway? Lambda functions can't have annotations on their parameters, so for lambdas the only things we need to analyse are the binding the parameter creates and the parameter's default value (if it has one). Both the binding and the default value seem to be taken care of outside the `visit_parameters()` call

Yes, technically there's no benefit of calling `visit_parameters` for lambdas here. I think it does make sense to keep it for consistency and in case the `Parameters` node is changed in the future by either us or CPython this will provide us some kind of failure (I think).

> For lambdas, they could. I think an alternate approach here would be to add `visit_parameters_without_defaults` and use that in the `FunctionDef` case (rather than overriding `visit_parameters`), and then remove the special code to visit defaults from the lambda case and just rely on `visit_parameters` here. I don't have a strong feeling that this is better, though -- the current approach more closely mirrors how type inference handles it. I do think it's nice to keep a parallel between the handling of lambda and FunctionDef, which the current code in the PR does.

I like the idea although I think I'd prefer consistency here.